Only get substrings in fingerprinting when supported [ci drivers]

dpsutton opened the pull request:
failing tests in bigquery and mongo as they don't support
expressions.


This regression was introduced with the OOM 2.1 fix but i didn't run the drivers tests unfortunately.

dpsutton committed:
Only get substrings in fingerprinting when supported [ci drivers]

failing tests in bigquery and mongo as they don't support
expressions.

sbelak commented on src/metabase/sync/analyze/fingerprint.clj:
```diff
@@ -48,25 +49,27 @@
-                        (empty-stats-map (count fingerprints))
-                        (map vector fields fingerprints))))
-             (metadata-queries/table-rows-sample table fields {:truncation-size truncation-size})))
+  (let [options (when (driver/supports? driver/*driver* :expressions)
```
I think a better place for this would be `table-rows-sample`. There's no need for `fingerprint-table!` to be aware of (and coupled to) implementation details and limitations of the sampling method it uses.

dpsutton replied:
yeah i went back and forth on that. wasn't sure who should be concerned with that. the caller asking for truncation or the implementation. felt a bit weird that you could ask for truncation at the call site but then not get it. felt like a bit of a surprise. but i'll move it

dpsutton replied:
gonna wait for this test pass to finish before pushing that up

sbelak replied:
That's a fair point. Pick whichever you think makes more sense.

dpsutton committed:
Corrects ns form and gets driver from table def [ci drivers]

getting driver from dynamic var was not reliable at this place

dpsutton committed:
Move expression check into `metadata-queries/table-row-sample`

debated whether caller should know that whether to pass in truncation
options or they might "silently" be ignored. Probably best this way
and added that information to docstring.

dpsutton commented:
@sbelak thanks so much for the help today. I think this is ready to go despite the four failing frontend tests. I am assuming that is general flakiness and can be ignored on this changeset.
